#### April 12 Class

_RNA-seq_

Workshop link on moodle

---

RNA-seq is sequencing all the RNA in what you decide to sequence

	Tissue type, cell
	
	Try to get mRNA - only 2% of the RNA
	
mRNA and Gene Expression

	Coding sequence -> pre-mRNA -splicing-> mRNA -> polypeptide 
	
Applications

	Differential gene expression between wild type and mutants
	
	Splice variants - easier if you have a reference genome
	
	Novel gene identification - transcripts that seem like something new
	
	SNP finding
	
	Transcriptome assembly
	
	RNA editing
	
Experimental Design

	What biological question?
	
	Samples? tissues? timepoints for the same tissue?
	
	How much sequence is needed?
	
	How long do these sequences need to be?
	
	Barcoding? Tag the individuals or tissue types? Pooling all samples together when sequencing? (cheaper sequencing)
	
	How many biological replicates? Or sample size/# of individuals?
	
	Technical replicates or how many sequencing lanes in Illumina lanes are needed? (replicates of sequencing reaction)	
Goal of experiment

	differences between wild-type and mutant?
	
	Generate hypotheses
	
Want about 30 million reads to get close to 100% of genes 

	Important if looking expression levels
	
Mostly done using Illumina

Total RNA -> mRNA -> Fragmented mRNA/cDNA -> Finished library (sequenced cDNA)

RNA Processing

	PolyA Selection - only mRNA has polyA tail
	
	rRNA Depletion - use RiboZero, RiboMinus to get rid of rRNA
	
Strand-Specific / Directional

	Important for alignments to genome
	
	Helpful in transcriptome assembly
	
Species to Choose From:

	ALUC
	EMBR
	ERWT
	IZNU
	NAUM
	NHAG
	OQBM
	QSLH	
	SDXI	
	VXKB
	
----	
	
Independent Project: NHAG		
	
Ipomoea nil (L.) Roth
whiteedge morning-glory 

Growth Habit: Forb/herb, Vine 

Distributed in California, Texas, Oklahoma, Louisiana, 			
	
	